### ISSUE TITLE
xunit.xml: FIXME: fill this with the [STDOUT|STDERR] of the test run

### ISSUE THREAD

**Issue by paweln007** [CiscoTestAutomation/pyats#199]
Passing the --xunit option to latest pyats to have a JUnit report generated:

pyats run job --xunit [job file]

The report does not have system-out and system-error entries populated.

```
<?xml version="1.0" encoding="UTF-8"?>
<testsuites disabled="0" errors="1" failures="0" tests="27" time="136.51647099999997">
        <testsuite disabled="0" errors="1" failures="0" name="1_smoke-suite" skipped="25" tests="27" time="136.51647099999997">
                <testcase name="001_common_setup" time="129.465092">
                        <error type="error" message="Test errored"> Please see the trade link to view the log: </error>
                        <system-out># FIXME: fill this with the STDOUT of the test run</system-out>
                        <system-err># FIXME: fill this with the STDERR of the test run</system-err>
                </testcase>
                <testcase name="002_NetworkInit" time="0.002119">
                        <skipped type="skipped" message="Test blocked"> Please see the trade link to view the log: </skipped>
                        <system-out># FIXME: fill this with the STDOUT of the test run</system-out>
                        <system-err># FIXME: fill this with the STDERR of the test run</system-err>
                </testcase>

```

**Comment by paweln007** [comment]
Another missing data piece is the text field of <error>, populated here with:   _Please see the trade link to view the log:_ 

`<error type="error" message="Test errored"> Please see the trade link to view the log: </error>`

**Comment by sumitsehgal123** [comment]
Hi ,
Please let me know you are looking for any support for your queries?

**Comment by paweln007** [comment]
`
$ pyats version check
You are currently running pyATS version: 22.1
Python: 3.10.0 [64bit]

  Package              Version
  -------------------- -------
  pyats                22.1   
  pyats.aereport       22.1   
  pyats.aetest         22.1   
  pyats.async          22.1   
  pyats.connections    22.1   
  pyats.datastructures 22.1   
  pyats.easypy         22.1   
  pyats.kleenex        22.1   
  pyats.log            22.1   
  pyats.reporter       22.1   
  pyats.results        22.1   
  pyats.tcl            22.1   
  pyats.topology       22.1   
  pyats.utils          22.1   
  unicon               22.1   
  unicon.plugins       22.1.1

$ cat env.txt
Runtime Values
==============
{'argv': {'all': '/usr/local/bin/pyats run job job.py --verbose --no-mail '
                 '--html-logs . --xunit .',
          'custom': ('',)},
 'host': {'architecture': 'x86_64',
          'distro': 'Fedora Linux 35 ',
          'kernel': '5.15.0-1031-azure',
          'name': 'jmp-qainfra-1'},
 'prefix': '/usr',
 'python': {'architecture': '64bit',
            'name': 'cpython',
            'tag': 'cpython-310',
            'version': '3.10.0'},
 'user': 'paweln'}
…

$ cat suite.py 
from pyats import aetest

class CommonSetup(aetest.CommonSetup):
    @aetest.subsection
    def connect_to_devices(self):
        pass

class Testcase0(aetest.Testcase):
    @aetest.test
    def test(self):
        pass

class Testcase1(aetest.Testcase):
    @aetest.test
    def test(self):
        assert 1 + 1 == 3

class Testcase2(aetest.Testcase):
    @aetest.test
    def test(self):
        self.failed('FAILED to veirfy xyz')

class Testcase3(aetest.Testcase):
    @aetest.test
    def test(self):
        self.errored('ERRORED while trying to veirfy xyz')

class CommonCleanup(aetest.CommonCleanup):
    @aetest.subsection
    def disconnect_from_devices(self):
        pass

if __name__ == '__main__':
    aetest.main()


$ cat xunit.xml 
<?xml version="1.0" encoding="UTF-8"?>
<testsuites disabled="0" errors="1" failures="2" tests="6" time="0.06748">
	<testsuite disabled="0" errors="1" failures="2" name="1_suite" skipped="0" tests="6" time="0.06748">
		<testcase name="001_common_setup" time="0.012023">
			<system-out># FIXME: fill this with the STDOUT of the test run</system-out>
			<system-err># FIXME: fill this with the STDERR of the test run</system-err>
		</testcase>
		<testcase name="002_Testcase0" time="0.011385">
			<system-out># FIXME: fill this with the STDOUT of the test run</system-out>
			<system-err># FIXME: fill this with the STDERR of the test run</system-err>
		</testcase>
		<testcase name="003_Testcase1" time="0.010427">
			<failure type="failure" message="Test failed"> Please see the trade link to view the log: </failure>
			<system-out># FIXME: fill this with the STDOUT of the test run</system-out>
			<system-err># FIXME: fill this with the STDERR of the test run</system-err>
		</testcase>
		<testcase name="004_Testcase2" time="0.012274">
			<failure type="failure" message="Test failed"> Please see the trade link to view the log: </failure>
			<system-out># FIXME: fill this with the STDOUT of the test run</system-out>
			<system-err># FIXME: fill this with the STDERR of the test run</system-err>
		</testcase>
		<testcase name="005_Testcase3" time="0.010987">
			<error type="error" message="Test errored"> Please see the trade link to view the log: </error>
			<system-out># FIXME: fill this with the STDOUT of the test run</system-out>
			<system-err># FIXME: fill this with the STDERR of the test run</system-err>
		</testcase>
		<testcase name="006_common_cleanup" time="0.010384">
			<system-out># FIXME: fill this with the STDOUT of the test run</system-out>
			<system-err># FIXME: fill this with the STDERR of the test run</system-err>
		</testcase>
	</testsuite>
</testsuites>
`

**Comment by sumitsehgal123** [comment]
Hi Paweln007,
I am working on the above issue and get back to you .

**Comment by sumitsehgal123** [comment]
Hi Hi Paweln007,
I seen the above issue, and I found that you are not connected to any actual devices  . So that the test cases is not able capture
stdout or sdterr  in initial  test xunit .Let me know if I can close this issue or you need any other help from our side .

**Comment by paweln007** [comment]
Here's a test suite which both prints on the stdout and connects to a device.  I used example mock test bed file and test script posted here:
https://pubhub.devnetcloud.com/media/pyats-getting-started/docs/_downloads/04c1c0ffd3a875e85db16c7408c0f784/mock.zip

The suite file:
$ cat suite.py 
from pyats import aetest
from genie.testbed import load

```
class Testcase(aetest.Testcase):

    @aetest.test
    def test(self):
        tb = load('mock.yaml')
        dev = tb.devices['nx-osv-1']
        dev.connect()
        p1 = dev.parse('show inventory')
        serial_number = p1['name']['Slot 1']['serial_number']
        print(f'My serial for slot1 is: {serial_number}')
        self.errored(f'Expected: 123456 but got: {serial_number}')

if __name__ == '__main__':
    aetest.main()
``` 

Command
`$ pyats run job job.py --xunit /tmp
`
Produces:
```
$ cat /tmp/xunit.xml 
<?xml version="1.0" encoding="UTF-8"?>
<testsuites disabled="0" errors="1" failures="0" tests="1" time="3.840489">
	<testsuite disabled="0" errors="1" failures="0" name="1_suite" skipped="0" tests="1" time="3.840489">
		<testcase name="001_Testcase" time="3.840489">
			<error type="error" message="Test errored"> Please see the trade link to view the log: </error>
			<system-out># FIXME: fill this with the STDOUT of the test run</system-out>
			<system-err># FIXME: fill this with the STDERR of the test run</system-err>
		</testcase>
	</testsuite>
</testsuites>
$

```
I am on pyats version 23.4

**Comment by sumitsehgal123** [comment]
Currently the development related discussion is going on with the team. Once it is done, I'll start working on it.

**Comment by sucharitha-emmanuel-westpac** [comment]
Is this issue resolved? I am facing the same issue so want to know the solution for the same.

**Comment by sucharitha-emmanuel-westpac** [comment]
@paweln007 did you get to know the solution for the above issue you have mentioned?

**Comment by sucharitha-emmanuel-westpac** [comment]
@paweln007 did you get to know the solution for the above issue you have mentioned?

**Comment by paweln007** [comment]
I have not received any updates on this item.

-Pawel

**Comment by sumitsehgal123** [comment]
Hi , 
I am working on your issue , and PR has been raised ,Working on unit test file .Once the PR has been merged and issue has been fix .I will informed you .

**Comment by davidwallis3101** [comment]
Any updates on this?

**Comment by SohanTirpude** [comment]
Hello @paweln007 and @davidwallis3101,

We added the fix for this issue and in the version of 24.2 this fix will be available. Meanwhile, each testcases' all stdout will be available in the `system-out` of each testcases in xunit.xml file.

For example, this is how it will look like:

```
<?xml version="1.0" encoding="UTF-8"?>
<testsuites disabled="0" errors="1" failures="0" tests="1" time="7.648091">
	<testsuite disabled="0" errors="1" failures="0" name="1_suite" skipped="0" tests="1" time="7.648091">
		<testcase name="001_Testcase" time="7.648091">
			<system-out>3: bgl-ads-6256: 2024-01-18T16:25:42: %AETEST-6-INFO: %[part=3.1/3][pid=2210693][pname=Task-1]: +------------------------------------------------------------------------------+
4: bgl-ads-6256: 2024-01-18T16:25:42: %AETEST-6-INFO: %[part=3.2/3][pid=2210693][pname=Task-1]: |                          Starting testcase Testcase                          |
5: bgl-ads-6256: 2024-01-18T16:25:42: %AETEST-6-INFO: %[part=3.3/3][pid=2210693][pname=Task-1]: +------------------------------------------------------------------------------+
6: bgl-ads-6256: 2024-01-18T16:25:42: %AETEST-6-INFO: %[part=6.1/3][pid=2210693][pname=Task-1]: +------------------------------------------------------------------------------+
7: bgl-ads-6256: 2024-01-18T16:25:42: %AETEST-6-INFO: %[part=6.2/3][pid=2210693][pname=Task-1]: |                            Starting section test                             |
8: bgl-ads-6256: 2024-01-18T16:25:42: %AETEST-6-INFO: %[part=6.3/3][pid=2210693][pname=Task-1]: +------------------------------------------------------------------------------+
9: bgl-ads-6256: 2024-01-18T16:25:42: %UNICON-6-INFO: %[pid=2210693][pname=Task-1]: +++ nx-osv-1 logfile /xxxx/users/x/runinfo/job.2024Jan18_16:25:35.571835/nx-osv-1-cli-1705575342.log +++
10: bgl-ads-6256: 2024-01-18T16:25:42: %UNICON-6-INFO: %[pid=2210693][pname=Task-1]: +++ Unicon plugin nxos (unicon.internal.plugins.nxos) +++
11: bgl-ads-6256: 2024-01-18T16:25:42: %WARNINGS-4-WARNING: %[part=11.1/3][pid=2210693][pname=Task-1]: /xxxx/lib/python3.8/site-packages/unicon/bases/routers/connection.py:97:
12: bgl-ads-6256: 2024-01-18T16:25:42: %WARNINGS-4-WARNING: %[part=11.2/3][pid=2210693][pname=Task-1]: DeprecationWarning: Arguments 'username', 'enable_password','tacacs_password' and 'line_password' are now deprecated and replaced by 'credentials'.
13: bgl-ads-6256: 2024-01-18T16:25:42: %WARNINGS-4-WARNING: %[part=11.3/3][pid=2210693][pname=Task-1]:   warnings.warn(message = &quot;Arguments 'username', &quot;
Trying mock_device ...
Connected to mock_device.
Escape character is '^]'.

15: bgl-ads-6256: 2024-01-18T16:25:45: %UNICON-6-INFO: %[pid=2210693][pname=Task-1]: +++ connection to spawn: mock_device_cli --os nxos --mock_data_dir mocked_data/nxos --state connect, id: 140703963845968 +++
16: bgl-ads-6256: 2024-01-18T16:25:45: %UNICON-6-INFO: %[pid=2210693][pname=Task-1]: connection to nx-osv-1

switch#
18: bgl-ads-6256: 2024-01-18T16:25:46: %UNICON-4-WARNING: %[pid=2210693][pname=Task-1]: Invalid hostname detected: found switch, expected nx-osv-1
19: bgl-ads-6256: 2024-01-18T16:25:46: %UNICON-6-INFO: %[pid=2210693][pname=Task-1]: +++ initializing handle +++
20: bgl-ads-6256: 2024-01-18T16:25:46: %UNICON-6-INFO: %[pid=2210693][pname=Task-1]: +++ nx-osv-1 with via 'ssh': executing command 'term length 0' +++
term length 0
switch#
22: bgl-ads-6256: 2024-01-18T16:25:46: %UNICON-6-INFO: %[pid=2210693][pname=Task-1]: +++ nx-osv-1 with via 'ssh': executing command 'term width 511' +++
term width 511
switch#
24: bgl-ads-6256: 2024-01-18T16:25:46: %UNICON-6-INFO: %[pid=2210693][pname=Task-1]: +++ nx-osv-1 with via 'ssh': executing command 'terminal session-timeout 0' +++
terminal session-timeout 0
switch#
26: bgl-ads-6256: 2024-01-18T16:25:46: %UNICON-6-INFO: %[pid=2210693][pname=Task-1]: +++ nx-osv-1(uut) with via 'ssh': configure +++
config term
switch(config)#no logging console
switch(config)#line console
switch(config)#exec-timeout 0
switch(config)#line vty
switch(config)#exec-timeout 0
switch(config)#terminal width 511
switch(config)#end
switch#
36: bgl-ads-6256: 2024-01-18T16:25:49: %UNICON-6-INFO: %[pid=2210693][pname=Task-1]: +++ nx-osv-1 with via 'ssh': executing command 'show inventory' +++
show inventory
NAME: &quot;Chassis&quot;,  DESCR: &quot;NX-OSv Chassis &quot;                       
PID: N7K-C7018           ,  VID: V00 ,  SN: TB00010000B          

NAME: &quot;Slot 1&quot;,  DESCR: &quot;NX-OSv Supervisor Module&quot;              
PID: N7K-SUP1            ,  VID: V00 ,  SN: TM00010000B          

NAME: &quot;Slot 2&quot;,  DESCR: &quot;NX-OSv Ethernet Module&quot;                
PID: N7K-F248XP-25       ,  VID: V00 ,  SN: TM00010000C          

NAME: &quot;Slot 3&quot;,  DESCR: &quot;NX-OSv Ethernet Module&quot;                
PID: N7K-F248XP-25       ,  VID: V00 ,  SN: TM00010000D          

NAME: &quot;Slot 4&quot;,  DESCR: &quot;NX-OSv Ethernet Module&quot;                
PID: N7K-F248XP-25       ,  VID: V00 ,  SN: TM00010000E          

NAME: &quot;Slot 33&quot;,  DESCR: &quot;NX-OSv Chassis Power Supply&quot;           
PID:                     ,  VID: V00 ,  SN:                      

NAME: &quot;Slot 35&quot;,  DESCR: &quot;NX-OSv Chassis Fan Module&quot;             
PID:                     ,  VID: V00 ,  SN:
switch#
38: bgl-ads-6256: 2024-01-18T16:25:49: %AETEST-3-ERROR: %[pid=2210693][pname=Task-1]: Errored reason: Expected: 123456 but got: TM00010000B
39: bgl-ads-6256: 2024-01-18T16:25:49: %AETEST-6-INFO: %[pid=2210693][pname=Task-1]: The result of section test is =&gt; ERRORED
40: bgl-ads-6256: 2024-01-18T16:25:49: %AETEST-6-INFO: %[pid=2210693][pname=Task-1]: The result of testcase Testcase is =&gt; ERRORED
41: bgl-ads-6256: 2024-01-18T16:25:4</system-out>
			<system-err># FIXME: fill this with the STDERR of the test run</system-err>
		</testcase>
	</testsuite>
</testsuites>
```

Kindly give it a try once pyATS 24.2 version is released. Meanwhile, I am closing this issue. Kindly feel free to re-open it if you have any issues related to it.

Thank you.
